File: cmsdev/internal/test/vcs/vcs.go
// MIT License
//
// (C) Copyright 2019-2023 Hewlett Packard Enterprise Development LP
//
// Permission is hereby granted, free of charge, to any person obtaining a
// copy of this software and associated documentation files (the "Software"),
// to deal in the Software without restriction, including without limitation
// the rights to use, copy, modify, merge, publish, distribute, sublicense,
// and/or sell copies of the Software, and to permit persons to whom the
// Software is furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included
// in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
// OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
// ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
// OTHER DEALINGS IN THE SOFTWARE.
package vcs

/*
 * vcs.go
 *
 * vcs commons file
 *
 */

import (
	"regexp"
	"stash.us.cray.com/SCMS/cms-tools/cmsdev/internal/lib/common"
	"stash.us.cray.com/SCMS/cms-tools/cmsdev/internal/lib/k8s"
	"stash.us.cray.com/SCMS/cms-tools/cmsdev/internal/lib/test"
	"strings"

	"k8s.io/apimachinery/pkg/apis/meta/v1"
)

type backupPod struct {
	Name    string
	Started v1.Time
	Status  string
}

func IsVCSRunning() (passed bool) {
	var latestBackupPod backupPod

	passed = true
	// We expect to find 2 or more gitea-vcs pods (one main pod, at least one postgres pod,
	// and one or more postgres backup pod)
	podNames, ok := test.GetPodNamesByPrefixKey("vcs", 2, -1)
	if !ok {
		passed = false
	}
	common.Infof("Found %d vcs pods: %s", len(podNames), strings.Join(podNames, ", "))

	pvcNames := make([]string, 0, 4)
	pvcNames = append(pvcNames, "gitea-vcs-data-claim")
	if !test.CheckPVCStatus(pvcNames[0]) {
		passed = false
	}
	// In addition to the main gitea-vcs pod (which should be in Running state), we expect:
	// - 1 or more gitea-vcs-postgres pod (in Running state)
	// - 0 or more gitea-vcs-wait-for-postgres-<num>- pods (in Succeeded state)
	// - 0 or more gitea-vcs-postgresql-db-backup- pods

	listedPodsOnError := false
	mainPodCount := 0
	postgresPodCount := 0
	waitForPostgresRe := regexp.MustCompile("gitea-vcs-wait-for-postgres-[0-9][0-9]*-")
	latestBackupPod.Name = ""
	for _, podName := range podNames {
		expectedStatus := "Running"
		if waitForPostgresRe.MatchString(podName) {
			expectedStatus = "Succeeded"
		} else if strings.HasPrefix(podName, "gitea-vcs-postgres-") {
			postgresPodCount += 1
			// There should be a corresponding pvc
			pvcName := "pgdata-" + podName
			common.Infof("There should be a corresponding pvc with name %s", pvcName)
			pvcNames = append(pvcNames, pvcName)
			if !test.CheckPVCStatus(pvcName) {
				passed = false
			}
		} else if strings.HasPrefix(podName, "gitea-vcs-postgresql-db-backup-") {
			common.Infof("checking pod start time for %s", podName)
			podStarted, err := k8s.GetPodStartTime(common.NAMESPACE, podName)
			if err != nil {
				common.VerboseFailedf(err.Error())
				passed = false
				continue
			}
			common.Infof("Pod start time is %v", podStarted)
			common.Infof("checking pod status for %s", podName)
			podStatus, err := k8s.GetPodStatus(common.NAMESPACE, podName)
			if err != nil {
				common.VerboseFailedf(err.Error())
				passed = false
				continue
			}
			if podStatus != "Succeeded" && podStatus != "Running" && podStatus != "Pending" {
				common.Warnf("Pod %s status is %s", podName, podStatus)
			} else {
				common.Infof("Pod %s status is %s", podName, podStatus)
			}
			if latestBackupPod.Name == "" || podStarted.After(latestBackupPod.Started.Time) {
				latestBackupPod.Name = podName
				latestBackupPod.Started = podStarted
				latestBackupPod.Status = podStatus
			}
			continue
		} else {
			mainPodCount += 1
		}
		common.Infof("checking pod status for %s expecting %s", podName, expectedStatus)
		status, err := k8s.GetPodStatus(common.NAMESPACE, podName)
		if err != nil {
			common.VerboseFailedf(err.Error())
			passed = false
			continue
		}
		common.Infof("Pod status is %s", status)
		if status != expectedStatus {
			if !listedPodsOnError {
				common.Printf("Found %d vcs pods: %s", len(podNames), strings.Join(podNames, ", "))
				listedPodsOnError = true
			}
			common.VerboseFailedf("expected status=%s, found status=%s for podName=%s", expectedStatus, status, podName)
			passed = false
		} else {
			common.VerboseOkay()
		}
	}
	if mainPodCount != 1 {
		if !listedPodsOnError {
			common.Printf("Found %d vcs pods: %s", len(podNames), strings.Join(podNames, ", "))
			listedPodsOnError = true
		}
		common.VerboseFailedf("Expected exactly 1 main gitea-vcs- pod but found %d", mainPodCount)
		passed = false
	}
	if postgresPodCount < 1 {
		if !listedPodsOnError {
			common.Printf("Found %d vcs pods: %s", len(podNames), strings.Join(podNames, ", "))
			listedPodsOnError = true
		}
		common.VerboseFailedf("Expected at least 1 gitea-vcs-postgres pod but found %d", postgresPodCount)
		passed = false
	}

	// Verify if postgres backup k8s cronjob exists
	err := k8s.VerifyCronJobExists(common.NAMESPACE, "gitea-vcs-postgresql-db-backup")
	if err == nil {
		common.Infof("kubernetes CronJob found in namespace %s with name %s", common.NAMESPACE, "gitea-vcs-postgresql-db-backup")
	} else {
		common.VerboseFailedf(err.Error())
		passed = false
	}

	if latestBackupPod.Name == "" {
		common.Warnf("No gitea-vcs-postgresql-db-backup pods found -- db has not yet been backed up")
	} else if latestBackupPod.Status == "Succeeded" {
		common.Infof("The latest gitea-vcs-postgresql-db backup pod (%s) completed successfully", latestBackupPod.Name)
	} else if latestBackupPod.Status == "Running" || latestBackupPod.Status == "Pending" {
		common.VerboseFailedf("The most recent gitea-vcs-postgresql-db-backup pod (%s) has not yet completed; re-run this test after it is done", latestBackupPod.Name)
	} else if latestBackupPod.Status == "Failed" {
		common.VerboseFailedf("The most recent gitea-vcs-postgresql-db-backup pod (%s) failed", latestBackupPod.Name)
		passed = false
	} else {
		common.VerboseFailedf("The most recent gitea-vcs-postgresql-db-backup pod (%s) has an unexpected status (%s)", latestBackupPod.Name, latestBackupPod.Status)
		passed = false
	}

	if !passed {
		common.ArtifactsKubernetes()
		if len(podNames) > 0 {
			common.ArtifactDescribeNamespacePods(common.NAMESPACE, podNames)
		}
		if len(pvcNames) > 0 {
			common.ArtifactDescribeNamespacePods(common.NAMESPACE, pvcNames)
		}
	}

	if !repoTest() {
		passed = false
	}

	return
}

File: ct-tests/lib/cms/vcs.py
# Copyright 2020-2021 Hewlett Packard Enterprise Development LP
#
# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included
# in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
# OTHER DEALINGS IN THE SOFTWARE.
#
# (MIT License)

"""
VCS-related test helper functions for CMS tests
"""

from .k8s import get_vcs_username_password
from .helpers import debug, info, raise_test_exception_error, run_cmd_list

def clone_vcs_repo(tmpdir):
    """
    Clone the config-management vcs repo into the specified temporary directory.
    Sets user.name and user.email for that repo.
    Returns the repo directory.
    """
    try:
        vcsuser, vcspass = get_vcs_username_password()
    except Exception as e:
        raise_test_exception_error(e, "to get vcs username and password from kubernetes")
    run_cmd_list(
        ["git", "clone", "https://%s:%s@api-gw-service-nmn.local/vcs/cray/config-management.git" % (vcsuser, vcspass)],
        cwd=tmpdir)
    git_repo_dir="%s/config-management" % tmpdir
    info("Cloned vcs repo to directory %s" % git_repo_dir)
    run_cmd_list(["git", "-C", git_repo_dir, "config", "user.email", "catfood@dogfood.mil"])
    debug("Set user.email for cloned repo")
    run_cmd_list(["git", "-C", git_repo_dir, "config", "user.name", "Rear Admiral Joseph Catfood"])
    debug("Set user.name for cloned repo")
    return git_repo_dir

def create_vcs_branch(repo_dir, branchname, base_branch="master"):
    """
    In the specified repo dir:
    1) checkout the base branch
    2) create and checkout a new branch with the specified name
    3) modify the motd yaml file to include the name of the branch
    4) Commit and push the change to vcs
    """
    git_motd_yaml="%s/roles/motd/defaults/main.yml" % repo_dir
    debug("Returning to master branch")
    run_cmd_list(["git", "checkout", base_branch], cwd=repo_dir)
    debug("Creating branch %s" % branchname)
    run_cmd_list(["git", "checkout", "-b", branchname], cwd=repo_dir)
    debug("Modify %s" % git_motd_yaml)
    with open(git_motd_yaml, "rt") as f:
        all_lines = f.read().splitlines()
    with open(git_motd_yaml, "wt+") as f:
        for line in all_lines[:-1]:
            f.write("%s\n" % line)
        f.write("%s branch=%s\n" % (all_lines[-1], branchname))
    debug("Add & commit change")
    run_cmd_list(["git", "commit", "-am", "CMS tests are the best"], cwd=repo_dir)
    debug("Push commit")
    run_cmd_list(["git", "push", "--set-upstream", "origin", branchname], cwd=repo_dir)

def remove_vcs_test_branches(repo_dir, test_vcs_branches):
    """
    Delete the specified branches from vcs
    """
    blist = list(test_vcs_branches)
    for branch in blist:
        debug("Deleting branch %s" % branch)
        run_cmd_list(["git", "push", "origin", "--delete", branch], cwd=repo_dir)
        test_vcs_branches.remove(branch)

